history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
> The discussion @masomel and I were having above was hinting at the better way though. Think about the end of each epoch as taking a snapshot of the current state of the tree. In your numbering, the order would be: 2, 2, 2, 2, then 3, 1, 4. That's how I redefined UpdateHistory above.

This way, new data would be inserted to the tree at the end of epoch. This is different from what is proposed in the paper (section 4.1.1)

It also rises another issue: we have 2 different STRs of each epoch, right? One from the beginning of the epoch to before the end of the epoch, and the other is at the end of the epoch. Did I missed something here?